README.rst
========
HotQueue
========
:Info: HotQueue is a Python library that allows you to use Redis as a message queue within your Python programs.
:Author: Richard Henry (http://github.com/richardhenry)

About HotQueue
==============

HotQueue is a Python library that allows you to use `Redis <http://code.google.com/p/redis/>`_ as a message queue within your Python programs.

The main advantage of this model is that there is no queue server to run, other than Redis. This is particularly ideal if you're already using Redis as a datastore elsewhere. To install it, run:

    easy_install -U hotqueue

The best place to get started is `the documentation <http://richardhenry.github.com/hotqueue/>`_.

The source code is available on `GitHub <http://github.com/richardhenry/hotqueue>`_.

To get help with HotQueue, use the `HotQueue Users mailing list
<http://groups.google.com/group/hotqueue-users>`_.

Contributing
============
The source is available on `GitHub <http://github.com/richardhenry/hotqueue>`_. To contribute to the project, fork it on GitHub and send a pull request, all contributions and suggestions are welcome.


sources/changelog.txt
=========
Changelog
=========

Changes in v0.2.0
=================
- Renamed ``dequeue()`` method on ``HotQueue`` to ``get()``
- Renamed ``enqueue()`` method on ``HotQueue`` to ``put()``
- Added ``HotQueue.worker()`` function decorator
- ``HotQueue.get()`` method now supports block and timeout arguments
- Added test suite

Changes in v0.1
===============
- Initial release

tests.py
# -*- coding: utf-8 -*-

"""Test suite for the HotQueue library. To run this test suite, execute this
Python program (``python tests.py``). Redis must be running on localhost:6379,
and a list key named 'hotqueue:testqueue' will be created and deleted in db 0
several times while the tests are running.
"""

from time import sleep
import threading
import unittest

from hotqueue import HotQueue


class HotQueueTestCase(unittest.TestCase):
    
    def setUp(self):
        """Create the queue instance before the test."""
        self.queue = HotQueue('testqueue')
    
    def tearDown(self):
        """Clear the queue after the test."""
        self.queue.clear()
    
    def test_consume(self):
        """Test the consume generator method."""
        nums = [1, 2, 3, 4, 5, 6, 7, 8]
        # Test blocking with timeout:
        self.queue.put(*nums)
        msgs = []
        for msg in self.queue.consume(timeout=1):
            msgs.append(msg)
        self.assertEquals(msgs, nums)
        # Test non-blocking:
        self.queue.put(*nums)
        msgs = []
        for msg in self.queue.consume(block=False):
            msgs.append(msg)
        self.assertEquals(msgs, nums)
    
    def test_cleared(self):
        """Test for correct behaviour if the Redis list does not exist."""
        self.assertEquals(len(self.queue), 0)
        self.assertEquals(self.queue.get(), None)
    
    def test_get_order(self):
        """Test that messages are get in the same order they are put."""
        alphabet = ['abc', 'def', 'ghi', 'jkl', 'mno']
        self.queue.put(alphabet[0], alphabet[1], alphabet[2])
        self.queue.put(alphabet[3])
        self.queue.put(alphabet[4])
        msgs = []
        msgs.append(self.queue.get())
        msgs.append(self.queue.get())
        msgs.append(self.queue.get())
        msgs.append(self.queue.get())
        msgs.append(self.queue.get())
        self.assertEquals(msgs, alphabet)
    
    def test_length(self):
        """Test that the length of a queue is returned correctly."""
        self.queue.put('a message')
        self.queue.put('another message')
        self.assertEquals(len(self.queue), 2)
    
    def test_worker(self):
        """Test the worker decorator."""
        colors = ['blue', 'green', 'red', 'pink', 'black']
        # Test blocking with timeout:
        self.queue.put(*colors)
        msgs = []
        @self.queue.worker(timeout=1)
        def appender(msg):
            msgs.append(msg)
        appender()
        self.assertEqual(msgs, colors)
        # Test non-blocking:
        self.queue.put(*colors)
        msgs = []
        @self.queue.worker(block=False)
        def appender(msg):
            msgs.append(msg)
        appender()
        self.assertEqual(msgs, colors)
    
    def test_threaded(self):
        """Threaded test of put and consume methods."""
        msgs = []
        def put():
            for num in range(3):
                self.queue.put('message %d' % num)
                sleep(0.1)
        def consume():
            for msg in self.queue.consume(timeout=1):
                msgs.append(msg)
        putter = threading.Thread(target=put)
        consumer = threading.Thread(target=consume)
        putter.start()
        consumer.start()
        for thread in [putter, consumer]:
            thread.join()
        self.assertEqual(msgs, ['message 0', 'message 1', 'message 2'])


if __name__ == "__main__":
    unittest.main()



hotqueue.py
# -*- coding: utf-8 -*-

"""HotQueue is a Python library that allows you to use Redis as a message queue
within your Python programs.
"""

from functools import wraps
import cPickle

from redis import Redis


__all__ = ['HotQueue']

__version__ = '0.2.0'


class HotQueue(object):
    
    """Simple FIFO message queue stored in a Redis list. Example:

    >>> from hotqueue import HotQueue
    >>> queue = HotQueue('myqueue', host='localhost', port=6379, db=0)
    
    :param name: name of the queue
    :param kwargs: additional kwargs to pass to :class:`Redis`, most commonly
        :attr:`host`, :attr:`port`, :attr:`db`
    """
    
    def __init__(self, name, **kwargs):
        self.name = name
        self.__redis = Redis(**kwargs)
    
    def __len__(self):
        return self.__redis.llen(self.key)
    
    def __repr__(self):
        return ('<HotQueue: \'%s\', host=\'%s\', port=%d, db=%d>' %
            (self.name, self.__redis.host, self.__redis.port, self.__redis.db))
    
    @property
    def key(self):
        """Return the key name used to store this queue in Redis, which is
        a concatenation of "hotqueue:" and :attr:`name`.
        """
        return 'hotqueue:%s' % self.name
    
    def clear(self):
        """Clear the queue of all messages, deleting the Redis key."""
        self.__redis.delete(self.key)
    
    def consume(self, **kwargs):
        """Return a generator that yields whenever a message is waiting in the
        queue. Will block otherwise. Example:

        >>> for msg in queue.consume(timeout=1):
        ...     print msg
        my message
        another message
        
        :param kwargs: any arguments that :meth:`~hotqueue.HotQueue.get` can
            accept (:attr:`block` will default to ``True`` if not given)
        """
        kwargs.setdefault('block', True)
        try:
            while True:
                msg = self.get(**kwargs)
                if msg is None:
                    break
                yield msg
        except KeyboardInterrupt:
            print; return
    
    def get(self, block=False, timeout=None):
        """Return a message from the queue. Example:
    
        >>> queue.get()
        'my message'
        >>> queue.get()
        'another message'
        
        :param block: whether or not to wait until a msg is available in
            the queue before returning; ``False`` by default
        :param timeout: when using :attr:`block`, if no msg is available
            for :attr:`timeout` in seconds, give up and return ``None``
        """
        if block:
            if timeout is None:
                timeout = 0
            msg = self.__redis.blpop(self.key, timeout=timeout)
            if msg is not None:
                msg = msg[1]
        else:
            msg = self.__redis.lpop(self.key)
        if msg is not None:
            msg = cPickle.loads(msg)
        return msg
    
    def put(self, *msgs):
        """Put one or more messages onto the queue. Example:
    
        >>> queue.put('my message')
        >>> queue.put('another message')
        """
        for msg in msgs:
            msg = cPickle.dumps(msg)
            self.__redis.rpush(self.key, msg)
    
    def worker(self, **kwargs):
        """Decorator for using a function as a queue worker. Example:
    
        >>> @queue.worker(timeout=1)
        ... def printer(msg):
        ...     print msg
        >>> printer()
        my message
        another message
        
        :param kwargs: any arguments that :meth:`~hotqueue.HotQueue.get` can
            accept (:attr:`block` will default to ``True`` if not given)
        """
        def decorator(worker):
            @wraps(worker)
            def wrapper():
                for msg in self.consume(**kwargs):
                    worker(msg)
            return wrapper
        return decorator



searchindex.js
Search.setIndex({objects:{"hotqueue.HotQueue":{consume:[1,0,1],get:[1,0,1],clear:[1,0,1],worker:[1,0,1],key:[1,2,1],put:[1,0,1]},hotqueue:{HotQueue:[1,1,1]}},terms:{"default":1,all:1,code:0,help:0,consum:[0,1,2],fifo:[1,2],particularli:0,follow:2,pull:2,paramet:1,hotqueu:[0,1,2,3],explain:2,elsewher:0,better:2,program:[0,2],mail:0,main:0,sourc:0,"return":[1,2],supervisord:2,get:[0,1,2,3],python:[0,2],safe:2,initi:3,now:[2,3],requir:0,introduct:2,like:2,changelog:[0,3],list:[0,1,2],yield:[1,2],server:[0,2],"try":2,item:[0,2],api:[0,1],timeout:[1,3],dequeu:3,manag:2,map:2,replac:2,some:2,second:[1,2],num:2,integ:2,pass:1,port:[1,2],librari:0,compat:0,databas:2,someth:2,brief:2,delet:1,version:0,print:[1,2],"import":[1,2],ever:2,method:3,localhost:[1,2],refer:[0,1],run:[0,2],kei:[1,2],gener:[1,2],here:2,host:[1,2],let:2,releas:3,wait:[1,2],would:2,befor:1,could:2,queue:[0,1,2],tutori:[0,2],def:[1,2],fals:1,mani:2,chang:[0,3],onto:[0,1,2],other:0,block:[1,2,3],decim:2,onc:2,within:0,easy_instal:0,put:[0,1,2,3],alreadi:0,wrap:2,instal:0,redi:[0,1,2],your:[0,2],given:1,from:[1,2],advantag:0,wai:2,support:3,two:2,messag:[0,1,2],name:[1,2],avail:[0,1],interest:2,suit:3,msg:1,addit:1,until:1,store:[1,2],more:[1,2],"function":[1,2,3],statement:2,worker:[0,1,2,3],ani:[1,2],myqueu:[1,2],github:0,enqueu:3,renam:3,"true":1,than:[0,2],concaten:1,none:1,save:2,whenev:[1,2],whether:1,work:2,see:2,asynchron:2,record:2,kwarg:1,can:[1,2],otherwis:[1,2],dec:2,loop:2,connect:[0,2],explor:0,creat:2,give:1,process:2,argument:[1,2,3],accept:1,minimum:0,have:2,option:2,decor:[1,2,3],built:2,receiv:2,anoth:[1,2],when:1,same:2,write:[0,2],how:2,ideal:0,instanc:2,which:[1,2],indefinit:2,test:3,you:[0,2],document:0,simpl:[1,2],even:2,printer:1,distribut:2,object:2,most:1,user:0,mai:2,datastor:0,wherev:2,"class":1,squar:2,task:2,off:[0,2],practic:2,clear:1,pickl:2,exampl:[1,2],allow:0,commonli:1,thi:[0,1,2],push:2,model:0,usual:2},objtypes:{"0":"py:method","1":"py:class","2":"py:attribute"},titles:["HotQueue User Documentation","API Reference","Tutorial","Changelog"],objnames:{"0":"Python method","1":"Python class","2":"Python attribute"},filenames:["index","apireference","tutorial","changelog"]})

hotqueue/__init__.py
# -*- coding: utf-8 -*-

import cPickle

from redis import Redis


__all__ = ['HotQueue']

__version__ = '0.1'


class HotQueue(object):
    
    """Simple FIFO message queue stored in a Redis list.
    
    :param name: name of the queue
    :param kwargs: additional kwargs to pass to :class:`Redis`, most commonly
        :attr:`host`, :attr:`port`, :attr:`db`
    """
    
    def __init__(self, name, **kwargs):
        """
        >>> queue = HotQueue('doctestq', host='localhost', port=6379, db=0)
        >>> queue.clear()
        >>> queue.enqueue('my message')
        >>> len(queue)
        1
        >>> queue.enqueue('another message')
        >>> len(queue)
        2
        >>> queue.dequeue()
        'my message'
        >>> queue.dequeue()
        'another message'
        >>> queue.clear()
        >>> queue.enqueue('dog')
        >>> len(queue)
        1
        >>> queue.enqueue('rabbit', 'mouse')
        >>> len(queue)
        3
        >>> for item in queue.consume(timeout=1):
        ...     print item
        dog
        rabbit
        mouse
        >>> queue.enqueue('Richard', 'Michael', 'Henry')
        >>> len(queue)
        3
        >>> queue.dequeue(block=True)
        'Richard'
        >>> len(queue)
        2
        >>> queue.dequeue(block=True, timeout=1)
        'Michael'
        >>> len(queue)
        1
        >>> for item in queue.consume(block=False):
        ...     print item
        Henry
        >>> len(queue)
        0
        >>> queue.enqueue({'cat': 'meow', 'dog': 'woof', 'cow': 'moo'})
        >>> sounds = queue.dequeue()
        >>> sounds['cat'] == 'meow'
        True
        >>> sounds['dog'] == 'woof'
        True
        >>> sounds['cow'] == 'moo'
        True
        >>> queue.clear()
        """
        self.name = name
        self.__redis = Redis(**kwargs)
    
    def __len__(self):
        return self.__redis.llen(self.key)
    
    def __repr__(self):
        return ('<HotQueue: \'%s\', host=\'%s\', port=%d, db=%d>' %
            (self.name, self.__redis.host, self.__redis.port, self.__redis.db))
    
    @property
    def key(self):
        """Return the key name used to store this queue in Redis, determined
        from :attr:`name`. Usually like ``hotqueue:name``.
        """
        return 'hotqueue:%s' % self.name
    
    def clear(self):
        """Clear the queue of all messages, deleting the Redis key."""
        self.__redis.delete(self.key)
    
    def consume(self, block=True, timeout=None):
        """Return a generator that yields whenever a message is waiting in the
        queue. Will block otherwise.
        
        :param block: whether or not to wait until a message is available in
            the queue before yielding; ``True`` by default
        :param timeout: when using :attr:`block`, if no message is available
            for :attr:`timeout` in seconds, return ``None``
        """
        try:
            while True:
                message = self.dequeue(block=block, timeout=timeout)
                if message is None:
                    break
                yield message
        except KeyboardInterrupt:
            print; return
    
    def dequeue(self, block=False, timeout=None):
        """Return a message from the queue.
        
        :param block: whether or not to wait until a message is available in
            the queue before returning; ``False`` by default
        :param timeout: when using :attr:`block`, if no message is available
            for :attr:`timeout` in seconds, return ``None``
        """
        if timeout is None:
            timeout = 0
        if block:
            message = self.__redis.blpop(self.key, timeout=timeout)
            if message is not None:
                message = message[1]
        else:
            message = self.__redis.lpop(self.key)
        if message is not None:
            message = cPickle.loads(message)
        return message
    
    def enqueue(self, *messages):
        """Push one or more messages onto the queue."""
        for message in messages:
            message = cPickle.dumps(message)
            self.__redis.rpush(self.key, message)

